### DISCUSSION TITLE
cleanup2: Cannot convert undefined or null to object

### DISCUSSION THREAD

**Discussion by thomas-0816** [alpinejs/alpine#3970] [7. Bugs (archived)]
Currently I'm implementing Tests with JSDOM in global context which run Alpine.start() multiple times.

I'm getting the following error:

```
TypeError: Cannot convert undefined or null to object
 ❯ ../src/node_modules/alpinejs/dist/module.esm.js:3041:12
    3039|   effect3(() => loop(el, iteratorNames, evaluateItems, evaluateKey));
    3040|   cleanup2(() => {
    3041|     Object.values(el._x_lookup).forEach((el2) => el2.remove());
       |            ^
    3042|     delete el._x_prevKeys;
    3043|     delete el._x_lookup;
 ❯ ../src/node_modules/alpinejs/dist/module.esm.js:720:49
 ❯ doCleanup ../src/node_modules/alpinejs/dist/module.esm.js:720:34
 ❯ ../src/node_modules/alpinejs/dist/module.esm.js:234:28
 ❯ ../src/node_modules/alpinejs/dist/module.esm.js:234:13
 ❯ cleanupAttributes ../src/node_modules/alpinejs/dist/module.esm.js:2
```

I'm suggesting to implement (similar to cleanupAttributes()):
```
  cleanup2(() => {
+    if (!el._x_lookup) return;
    Object.values(el._x_lookup).forEach((el2) => el2.remove());
    delete el._x_prevKeys;
    delete el._x_lookup;
  });
```

Also it would be nice to have a Alpine.stop() method next to Alpine.start() doing cleanup, stopping MutationObservers, etc.

version: "3.13.3",
node v20.10.0

**Comment by ekwoka** [comment]
I am working on a testing library for vitest that will do this all with happy-dom.

What you want to do to solve is isolate the tests.

What version of JSDOM? I assume the latest?

are there any other issues with elements having these internal props added?

**Reply by thomas-0816** [reply]
No other issues with JSDOM, code is here:
https://github.com/thbley/php_frameworkless/blob/master/tasks_fe/tests/ssr/utils/jsdom.js

Note: I'm using the system loader with ES modules to get code coverage, so I need to go global:
```
    const { Bootstrap } = await import(script);
    const Alpine = (new Bootstrap()).start();
```

----

Code for happy-dom is here (10-15% faster, but no support for document.evaluate(xpath)):
https://github.com/thbley/php_frameworkless/blob/master/tasks_fe/tests/ssr/utils/happydom.js

With happy-dom I have another issue:
```
TypeError: i is not a function
 ❯ ../src/node_modules/alpinejs/dist/module.esm.js:1072:28
    1070|             el2._x_hidePromise,
    1071|             ...(el2._x_hideChildren || []).map(hideAfterChildren)
    1072|           ]).then(([i]) => i());
       |                            ^
    1073|           delete el2._x_hidePromise;
    1074|           delete el2._x_hideChildren;
```

I patched this with:
```
- ]).then(([i]) => i());
+ ]).then(([i]) => i && i());
```

jsdom: 23.2.0
happy-dom: 13.0.0
vitest: 1.1.3

vitest.config.js
```
import { defineConfig } from 'vitest/config';

export default defineConfig({
    test: {
        reporters: ['verbose'],
        cacheDir: '/tmp/',
        logHeapUsage: true,
        coverage: {
            enabled: true,
            allowExternal: true,
            provider: 'v8',
            reporter: [['text', { maxCols: 120 }]],
            exclude: ['**/tests/**']
        }
    }
});
```

vitest-test: https://github.com/thbley/php_frameworkless/blob/master/tasks_fe/tests/ssr/index-html.test.js
